File: minimum_window_substring.cpp
/*
76. Minimum Window Substring

Time Complexity: O(m + n) where m = len(s), n = len(t)
Space Complexity: O(m + n) for the hashmaps
*/

#include <iostream>
#include <string>
#include <unordered_map>
#include <climits>
using namespace std;


class Solution {
public:
    string minWindow(string s, string t) {        
        if (s.empty() || t.empty() || s.length() < t.length()) {
            return "";
        }
        unordered_map<char, int> tFreq;
        for (char c : t) {
            tFreq[c]++;
        }
        int required = tFreq.size();
        unordered_map<char, int> windowFreq;
        int formed = 0; 
        int left = 0;
        int minLen = INT_MAX;
        int minLeft = 0;
        
        for (int right = 0; right < s.length(); right++) {
            char c = s[right];
            windowFreq[c]++;
            
            if (tFreq.find(c) != tFreq.end() && windowFreq[c] == tFreq[c]) {
                formed++;
            }
            while (formed == required && left <= right) {
                if (right - left + 1 < minLen) {
                    minLen = right - left + 1;
                    minLeft = left;
                }
                char leftChar = s[left];
                windowFreq[leftChar]--;
                if (tFreq.find(leftChar) != tFreq.end() && windowFreq[leftChar] < tFreq[leftChar]) {
                    formed--;
                }
                left++;
            }
        }
        
        return minLen == INT_MAX ? "" : s.substr(minLeft, minLen);
    }
};


// Test cases
int main() {
    Solution solution;
    
    // Example 1
    cout << "Example 1: '" << solution.minWindow("ADOBECODEBANC", "ABC") << "'" << endl;  // Expected: "BANC"
    
    // Example 2
    cout << "Example 2: '" << solution.minWindow("a", "a") << "'" << endl;  // Expected: "a"
    
    // Example 3
    cout << "Example 3: '" << solution.minWindow("a", "aa") << "'" << endl;  // Expected: ""
    
    // Edge case: t longer than s
    cout << "Example 4: '" << solution.minWindow("ab", "abc") << "'" << endl;  // Expected: ""
    
    // Edge case: exact match
    cout << "Example 5: '" << solution.minWindow("abc", "abc") << "'" << endl;  // Expected: "abc"
    
    // Edge case: multiple windows
    cout << "Example 6: '" << solution.minWindow("aaabbbccc", "abc") << "'" << endl;  // Expected: "abbc" or similar
    
    // Edge case: with duplicates
    cout << "Example 7: '" << solution.minWindow("ADOBECODEBANC", "AABC") << "'" << endl;  // Expected: "ADOBEC" or similar
    
    // Edge case: t has duplicate characters
    cout << "Example 8: '" << solution.minWindow("aabbcc", "abc") << "'" << endl;  // Expected: "abbc" or similar
    
    // Edge case: no common characters
    cout << "Example 9: '" << solution.minWindow("abc", "xyz") << "'" << endl;  // Expected: ""
    
    // Edge case: entire string is minimum
    cout << "Example 10: '" << solution.minWindow("abc", "cba") << "'" << endl;  // Expected: "abc"
    
    return 0;
}

File: minimum_window_substring.py
"""
76. Minimum Window Substring

Time Complexity: O(m + n) where m = len(s), n = len(t)
Space Complexity: O(m + n) for the hashmaps
"""

from collections import Counter, defaultdict


class Solution:
    def minWindow(self, s: str, t: str) -> str:
        if not s or not t or len(s) < len(t):
            return ""
        
        t_freq = Counter(t)
        required = len(t_freq) 
        window_freq = defaultdict(int)
        formed = 0 
        left = 0
        min_len = float('inf')
        min_left = 0
        
        for right in range(len(s)):
            char = s[right]
            window_freq[char] += 1
            if char in t_freq and window_freq[char] == t_freq[char]:
                formed += 1
            while formed == required and left <= right:
                if right - left + 1 < min_len:
                    min_len = right - left + 1
                    min_left = left
                char = s[left]
                window_freq[char] -= 1
                if char in t_freq and window_freq[char] < t_freq[char]:
                    formed -= 1
                
                left += 1
        
        return "" if min_len == float('inf') else s[min_left:min_left + min_len]
 
# Test cases
if __name__ == "__main__":
    solution = Solution()
    
    # Example 1
    s1 = "ADOBECODEBANC"
    t1 = "ABC"
    print(f"Example 1: '{solution.minWindow(s1, t1)}'")  # Expected: "BANC"
    
    # Example 2
    s2 = "a"
    t2 = "a"
    print(f"Example 2: '{solution.minWindow(s2, t2)}'")  # Expected: "a"
    
    # Example 3
    s3 = "a"
    t3 = "aa"
    print(f"Example 3: '{solution.minWindow(s3, t3)}'")  # Expected: ""
    
    # Edge case: t longer than s
    s4 = "ab"
    t4 = "abc"
    print(f"Example 4: '{solution.minWindow(s4, t4)}'")  # Expected: ""
    
    # Edge case: exact match
    s5 = "abc"
    t5 = "abc"
    print(f"Example 5: '{solution.minWindow(s5, t5)}'")  # Expected: "abc"
    
    # Edge case: multiple windows
    s6 = "aaabbbccc"
    t6 = "abc"
    print(f"Example 6: '{solution.minWindow(s6, t6)}'")  # Expected: "abbc" or "bbc"
    
    # Edge case: with duplicates
    s7 = "ADOBECODEBANC"
    t7 = "AABC"
    print(f"Example 7: '{solution.minWindow(s7, t7)}'")  # Expected: "ADOBEC" or "ODEBANC"
    
    # Edge case: t has duplicate characters
    s8 = "aabbcc"
    t8 = "abc"
    print(f"Example 8: '{solution.minWindow(s8, t8)}'")  # Expected: "abbc" or similar
    
    # Edge case: no common characters
    s9 = "abc"
    t9 = "xyz"
    print(f"Example 9: '{solution.minWindow(s9, t9)}'")  # Expected: ""
    
    # Edge case: entire string is minimum
    s10 = "abc"
    t10 = "cba"
    print(f"Example 10: '{solution.minWindow(s10, t10)}'")  # Expected: "abc"